src/benchmark/base_charm.py

```diff
@@ -0,0 +1,326 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""This class implements the default benchmark workflow.
+
+It is a functional charm that connects the benchmark service to the database and the grafana agent.
+
+The charm should inherit from this class and implement only the specifics for its own tool.
+
+The main step is to execute the run action. This action renders the systemd service file and
+starts the service. If the target is missing, then service errors and returns an error to
+the user.
+
+This charm should also be the main entry point to all the modelling of your benchmark tool.
+"""
+
+import logging
+import subprocess
+from abc import ABC, abstractmethod
+from typing import Any
+
+import ops
+from charms.grafana_agent.v0.cos_agent import COSAgentProvider
+from ops.charm import CharmEvents
+from ops.framework import EventBase, EventSource
+from ops.model import BlockedStatus
+
+from benchmark.core.models import DPBenchmarkLifecycleState
+from benchmark.core.pebble_workload_base import DPBenchmarkPebbleWorkloadBase
+from benchmark.core.systemd_workload_base import DPBenchmarkSystemdWorkloadBase
+from benchmark.core.workload_base import WorkloadBase
+from benchmark.events.db import DatabaseRelationHandler
+from benchmark.events.peer import PeerRelationHandler
+from benchmark.literals import (
+    COS_AGENT_RELATION,
+    METRICS_PORT,
+    PEER_RELATION,
+    DPBenchmarkLifecycleTransition,
+    DPBenchmarkMissingOptionsError,
+)
+from benchmark.managers.config import ConfigManager
+from benchmark.managers.lifecycle import LifecycleManager
+
+# Log messages can be retrieved using juju debug-log
+logger = logging.getLogger(__name__)
+
+
+class DPBenchmarkCheckUploadEvent(EventBase):
+    """Informs to check upload is finished."""
+
+
+class DPBenchmarkCheckCollectEvent(EventBase):
+    """Informs to check collect is finished."""
+
+
+class DPBenchmarkEvents(CharmEvents):
+    """Events used by the charm to check the upload."""
+
+    check_collect = EventSource(DPBenchmarkCheckCollectEvent)
+    check_upload = EventSource(DPBenchmarkCheckUploadEvent)
+
+
+def workload_build(workload_params_template: str) -> WorkloadBase:
+    """Build the workload."""
+    try:
+        # Really simple check to see if we have systemd
+        subprocess.check_output(["systemctl", "--help"])
+    except subprocess.CalledProcessError:
+        return DPBenchmarkPebbleWorkloadBase(workload_params_template)
+    return DPBenchmarkSystemdWorkloadBase(workload_params_template)
+
+
+class DPBenchmarkCharmBase(ops.CharmBase, ABC):
```

Comment:
**question** why aren't we using the handler pattern here and prefer instead charm base abstractions? In seems that here we are defining a bunch of hooks and registering them, as in handlers. To be clear, I'm not opposing this if we have good reasons, hence wondering what prompted you here